typing/rec_check.ml

```diff
@@ -1202,17 +1205,16 @@ and is_destructuring_pattern : type k . k general_pattern -> bool =
         is_destructuring_pattern l || is_destructuring_pattern r
 
 let is_valid_recursive_expression idlist expr =
-  let ty = expression expr Return in
-  match Env.unguarded ty idlist, Env.dependent ty idlist,
-        classify_expression expr with
-  | _ :: _, _, _ (* The expression inspects rec-bound variables *)
-  | [], _ :: _, Dynamic -> (* The expression depends on rec-bound variables
-                              and its size is unknown *)
-      false
-  | [], _, Static (* The expression has known size *)
-  | [], [], Dynamic -> (* The expression has unknown size,
-                          but does not depend on rec-bound variables *)
-      true
+  match classify_expression expr with
+  | Static_Delay -> true
```

Comment:
I'm not fond of this approach to implement the `Rec_check` fast path, can you explain why you did things that way?

In my mind there are two independent computations regarding recursive values:
- their "size"
- their "modes"

The final part of this module, that is the function here, combines the two analyses to make the final accept/reject decision. 

Your change introduces a new notion of size `Static_Delay` whose name mentions a mode and which means "fast path please", and is used for expressions of a certain shape. But here (in the "final" part) we do have the expression at hand, so why don't we check for functions here as a first check (and other fast-path conditions if any), then ask for the size in the non-function case, and then the modes?

(Remark: your change also contains a small refactoring where we check for the size first and then do the mode-checking differently, it does make the code nicer and it's a good approach to keep. But I have the impression that we could have this without the weird  `Static_Delay` size.)

Reply:
The old code calls `classify_expression expr` (cheap) and `expression expr Return` (expensive), and case-splits on the results. The observation here is that if `classify_expression expr` classified into three rather than two cases, you could skip the expensive call to `expression expr Return` in the common case, because functions have both static size and guarantee to delay all accesses.

I have no particular opinions on how this should be done, I'm not at all familiar with this code. I'm happy with any approach as long as it avoids calling `expression expr Return` on functions.